File: packages/typescript/src/accessibility/ChromiumAccessibilityTree.ts
import { always } from "alwaysly";
import { Element } from "domhandler";
import { Telemetry } from "../telemetry/Telemetry.ts";
import { Xml } from "../xml/Xml.ts";
import { XmlRenderer } from "../xml/XmlRenderer.ts";
import type { AccessibilityElement } from "./AccessibilityElement.ts";
import { BaseAccessibilityTree } from "./BaseAccessibilityTree.ts";

const { tracer } = Telemetry.get(import.meta.url);
const { span } = tracer.dec();
const PRESERVE_FALSE_ATTRS = new Set(["checked"]);

interface CDPNode {
  nodeId?: string | number;
  parentId?: string | number | null;
  backendDOMNodeId?: number;
  role?: { value?: string };
  name?: { value?: string };
  ignored?: boolean;
  properties?: Array<{
    name?: string;
    value?: unknown;
  }>;
  childIds?: Array<string | number>;
  _frame?: object;
  _parent_iframe_backend_node_id?: number;
  _frame_chain?: number[];
}

export class ChromiumAccessibilityTree extends BaseAccessibilityTree<
  Record<string, unknown> | string
> {
  #cdpResponse: Record<string, unknown>;
  // TODO: There's a bug in Bun that results in `#nextRawId` compiled to
  // `__privateGet(this, _nextRawId)++;` which causes a runtime error.
  // Figure out a solution to use private fields without breaking Bun
  // compatibility.
  private nextRawId: number = 0;
  #frameMap: Record<number, object> = {}; // raw_id -> Frame object for iframe support
  #frameChainMap: Record<number, number[]> = {}; // raw_id -> frame chain (list of iframe backendNodeIds)

  protected override get kind(): string {
    return "chromium";
  }

  constructor(
    cdpResponseOrXml: Record<string, unknown> | string,
    frameMap?: Record<number, object>,
  ) {
    super(cdpResponseOrXml);

    if (typeof cdpResponseOrXml === "object") {
      this.#cdpResponse = cdpResponseOrXml;
      return;
    }

    this.#cdpResponse = {};
    this.xml = cdpResponseOrXml;
    if (frameMap) this.#frameMap = frameMap;
  }

  /** Convert CDP response to raw XML format preserving all data. */
  @span("driver.tree.to_str", { "driver.tree.platform": "chromium" })
  toStr(): string {
    if (this.xml !== null) {
      return this.xml;
    }

    const nodes = (this.#cdpResponse.nodes as CDPNode[] | undefined) ?? [];
    if (nodes.length === 0) {
      return (this.xml = "");
    }

    // Create a lookup table for nodes by their ID
    const nodeLookup: Record<string, CDPNode> = {};
    for (const node of nodes) {
      if (node.nodeId !== undefined) {
        nodeLookup[String(node.nodeId)] = node;
      }
    }

    // Build mapping: backendDOMNodeId -> list of iframe child root nodes
    // This allows us to inline iframe content inside their parent <Iframe> elements
    const iframeChildren: Record<number, CDPNode[]> = {};
    const trueRoots: CDPNode[] = [];

    for (const node of nodes) {
      if (!node.parentId) {
        const parentIframeId = node._parent_iframe_backend_node_id;
        if (parentIframeId) {
          iframeChildren[parentIframeId] ??= [];
          iframeChildren[parentIframeId].push(node);
        } else {
          trueRoots.push(node);
        }
      }
    }

    // Build tree structure and convert to XML (only from true roots)
    const rootNodes: Element[] = [];
    for (const node of trueRoots) {
      const xmlNode = this.#nodeToXml(node, nodeLookup, iframeChildren);
      rootNodes.push(xmlNode);
    }

    // Combine all root nodes into a single XML string
    let xmlString = "";
    for (const root of rootNodes) {
      xmlString += XmlRenderer.render([root], {
        preserveFalseAttrs: PRESERVE_FALSE_ATTRS,
      });
    }

    this.xml = xmlString;
    return this.xml;
  }

  /** Convert a CDP node to XML element, recursively processing children. */
  #nodeToXml(
    node: CDPNode,
    nodeLookup: Record<string, CDPNode>,
    iframeChildren: Record<number, CDPNode[]>,
  ): Element {
    // Create element with role as tag
    const role = node.role?.value ?? "unknown";
    const elem = new Element(role, {});

    // Add our own sequential raw_id attribute
    this.nextRawId++;
    elem.attribs["raw_id"] = String(this.nextRawId);

    // Store frame reference if present (for iframe support)
    if ("_frame" in node && node._frame) {
      this.#frameMap[this.nextRawId] = node._frame;
    }

    // Store frame chain if present (for Selenium nested frame switching)
    if ("_frame_chain" in node && node._frame_chain) {
      this.#frameChainMap[this.nextRawId] = node._frame_chain;
    }

    // Add all node attributes as XML attributes
    if ("backendDOMNodeId" in node && node.backendDOMNodeId !== undefined) {
      elem.attribs["backendDOMNodeId"] = String(node.backendDOMNodeId);
    }
    if ("nodeId" in node && node.nodeId !== undefined) {
      elem.attribs["nodeId"] = String(node.nodeId);
    }
    if ("ignored" in node && node.ignored !== undefined) {
      elem.attribs["ignored"] = String(node.ignored);
    }

    // Add name as attribute if present
    if ("name" in node && node.name && "value" in node.name) {
      elem.attribs["name"] = String(node.name.value);
    }

    // Add value as attribute if present
    if (
      "value" in node &&
      typeof node.value === "object" &&
      node.value &&
      "value" in node.value
    ) {
      elem.attribs["value"] = String(node.value.value);
    }

    // Add properties as attributes
    for (const prop of node.properties || []) {
      const propName = prop.name ?? "";
      const propValue = prop.value ?? {};
      if (
        typeof propValue === "object" &&
        propValue !== null &&
        "value" in (propValue as Record<string, unknown>)
      ) {
        elem.attribs[propName] = String(
          (propValue as { value?: unknown }).value,
        );
      } else if (typeof propValue === "object" && propValue !== null) {
        // Complex property values (like nodeList) are converted to empty string
        elem.attribs[propName] = "";
      } else {
        elem.attribs[propName] = String(propValue);
      }
    }

    // Process children recursively
    if (node.childIds) {
      for (const childIdAny of node.childIds) {
        const childId = String(childIdAny);
        if (nodeLookup[childId]) {
          const childElem = this.#nodeToXml(
            nodeLookup[childId],
            nodeLookup,
            iframeChildren,
          );
          childElem.parent = elem;
          if (elem.children.length > 0) {
            const prev = elem.children[elem.children.length - 1];
            always(prev);
            prev.next = childElem;
            childElem.prev = prev;
          }
          elem.children.push(childElem);
        }
      }
    }

    // Inline iframe content: if this element is an iframe, add its child trees
    const backendNodeId = node.backendDOMNodeId;
    const iframeRoots = backendNodeId
      ? iframeChildren[backendNodeId]
      : undefined;
    if (backendNodeId && iframeRoots) {
      delete iframeChildren[backendNodeId];
      for (const childRoot of iframeRoots) {
        const childElem = this.#nodeToXml(
          childRoot,
          nodeLookup,
          iframeChildren,
        );
        childElem.parent = elem;
        if (elem.children.length > 0) {
          const prev = elem.children[elem.children.length - 1];
          always(prev);
          prev.next = childElem;
          childElem.prev = prev;
        }
        elem.children.push(childElem);
      }
    }

    return elem;
  }

  /**
   * Find element by raw_id and return its properties for element finding.
   *
   * @param rawId The raw_id to search for
   * @returns AccessibilityElement with backend_node_id set
   */
  @span("driver.tree.element_by_id", { "driver.tree.platform": "chromium" })
  elementById(rawId: number): AccessibilityElement {
    // Get raw XML with raw_id attributes
    const rawXml = this.toStr();
    const root = Xml.parseRoot(`<root>${rawXml}</root>`);
    // Find element with matching raw_id
    const findElement = (elem: Element, targetId: string): Element | null => {
      if (elem.attribs["raw_id"] === targetId) {
        return elem;
      }
      for (const child of Array.from(elem.children)) {
        const childEl = Xml.nodeAsTag(child);
        if (!childEl) {
          continue;
        }
        const result = findElement(childEl, targetId);
        if (result !== null) {
          return result;
        }
      }
      return null;
    };

    const element = findElement(root, String(rawId));
    if (element === null) {
      throw new Error(`No element with raw_id=${rawId} found`);
    }

    // Extract backendDOMNodeId for Chromium CDP nodes
    const backendNodeIdStr = element.attribs["backendDOMNodeId"];
    if (backendNodeIdStr === undefined) {
      throw new Error(
        `Element with raw_id=${rawId} has no backendDOMNodeId attribute`,
      );
    }

    return {
      type: element.tagName,
      backendNodeId: parseInt(backendNodeIdStr),
      frame: this.#frameMap[rawId],
      frameChain: this.#frameChainMap[rawId],
    };
  }

  /** Scope the tree to a smaller subtree identified by raw_id. */
  @span("driver.tree.scope_to_area", { "driver.tree.platform": "chromium" })
  scopeToArea(rawId: number): ChromiumAccessibilityTree {
    const rawXml = this.toStr();

    // Parse the XML
    const root = Xml.parseRoot(`<root>${rawXml}</root>`);

    // Find the element with the matching raw_id
    const findElement = (elem: Element, targetId: string): Element | null => {
      if (elem.attribs["raw_id"] === targetId) {
        return elem;
      }
      for (const child of Array.from(elem.children)) {
        const childEl = Xml.nodeAsTag(child);
        if (!childEl) {
          continue;
        }
        const result = findElement(childEl, targetId);
        if (result !== null) {
          return result;
        }
      }
      return null;
    };

    const targetElem = findElement(root, String(rawId));

    if (targetElem === null) {
      // If not found, return original tree
      return this;
    }

    // Convert the scoped element back to XML string
    const scopedXml = XmlRenderer.render([targetElem], {
      preserveFalseAttrs: PRESERVE_FALSE_ATTRS,
    });

    return new ChromiumAccessibilityTree(scopedXml, this.#frameMap);
  }
}

File: packages/python/src/alumnium/accessibility/chromium_accessibility_tree.py
from xml.etree.ElementTree import Element, fromstring, indent, tostring

from .accessibility_element import AccessibilityElement
from .base_accessibility_tree import BaseAccessibilityTree


class ChromiumAccessibilityTree(BaseAccessibilityTree):
    def __init__(self, cdp_response: dict):
        self.cdp_response = cdp_response
        self._next_raw_id = 0
        self._raw = None
        self._frame_map: dict[int, object] = {}  # raw_id -> Frame object for iframe support
        self._frame_chain_map: dict[int, list[int]] = {}  # raw_id -> frame chain (list of iframe backendNodeIds)

    @classmethod
    def _from_xml(cls, xml_string: str, frame_map: dict[int, object] | None = None) -> "ChromiumAccessibilityTree":
        """Create a ChromiumAccessibilityTree instance from pre-computed XML."""
        instance = cls(cdp_response={})
        instance._raw = xml_string
        if frame_map:
            instance._frame_map = frame_map
        return instance

    def to_str(self) -> str:
        """Convert CDP response to raw XML format preserving all data."""
        if self._raw is not None:
            return self._raw

        nodes = self.cdp_response.get("nodes", [])
        if not nodes:
            self._raw = ""
            return self._raw

        # Create a lookup table for nodes by their ID
        node_lookup = {node["nodeId"]: node for node in nodes}

        # Build mapping: backendDOMNodeId -> list of iframe child root nodes
        # This allows us to inline iframe content inside their parent <Iframe> elements
        iframe_children: dict[int, list[dict]] = {}
        true_roots: list[dict] = []

        for node in nodes:
            if node.get("parentId") is None:
                parent_iframe_id = node.get("_parent_iframe_backend_node_id")
                if parent_iframe_id:
                    iframe_children.setdefault(parent_iframe_id, []).append(node)
                else:
                    true_roots.append(node)

        # Build tree structure and convert to XML (only from true roots)
        root_nodes = []
        for node in true_roots:
            xml_node = self._node_to_xml(node, node_lookup, iframe_children)
            root_nodes.append(xml_node)

        # Combine all root nodes into a single XML string
        xml_string = ""
        for root in root_nodes:
            indent(root)
            xml_string += tostring(root, encoding="unicode")

        self._raw = xml_string
        return self._raw

    def _node_to_xml(self, node: dict, node_lookup: dict, iframe_children: dict[int, list[dict]]) -> Element:
        """Convert a CDP node to XML element, recursively processing children."""
        # Create element with role as tag
        role = node.get("role", {}).get("value", "unknown")
        elem = Element(role)

        # Add our own sequential raw_id attribute
        self._next_raw_id += 1
        elem.set("raw_id", str(self._next_raw_id))

        # Store frame reference if present (for iframe support)
        if "_frame" in node:
            self._frame_map[self._next_raw_id] = node["_frame"]

        # Store frame chain if present (for Selenium nested frame switching)
        if "_frame_chain" in node:
            self._frame_chain_map[self._next_raw_id] = node["_frame_chain"]

        # Add all node attributes as XML attributes
        if "backendDOMNodeId" in node:
            elem.set("backendDOMNodeId", str(node["backendDOMNodeId"]))
        if "nodeId" in node:
            elem.set("nodeId", str(node["nodeId"]))
        if "ignored" in node:
            elem.set("ignored", self._to_str(node["ignored"]))

        # Add name as attribute if present
        if "name" in node and "value" in node["name"]:
            elem.set("name", node["name"]["value"])

        # Add value as attribute if present
        if "value" in node and isinstance(node["value"], dict) and "value" in node["value"]:
            elem.set("value", self._to_str(node["value"]["value"]))

        # Add properties as attributes
        if "properties" in node:
            for prop in node["properties"]:
                prop_name = prop.get("name", "")
                prop_value = prop.get("value", {})
                if isinstance(prop_value, dict) and "value" in prop_value:
                    elem.set(prop_name, self._to_str(prop_value["value"]))
                elif isinstance(prop_value, dict):
                    # Complex property values (like nodeList) are converted to empty string
                    elem.set(prop_name, "")
                else:
                    elem.set(prop_name, self._to_str(prop_value))

        # Process children recursively
        if "childIds" in node:
            for child_id in node["childIds"]:
                if child_id in node_lookup:
                    child_elem = self._node_to_xml(node_lookup[child_id], node_lookup, iframe_children)
                    elem.append(child_elem)

        # Inline iframe content: if this element is an iframe, add its child trees
        backend_node_id = node.get("backendDOMNodeId")
        if backend_node_id and backend_node_id in iframe_children:
            for child_root in iframe_children.pop(backend_node_id):
                child_elem = self._node_to_xml(child_root, node_lookup, iframe_children)
                elem.append(child_elem)

        return elem

    def _to_str(self, value) -> str:
        """Convert a value to string, normalizing booleans to lowercase."""
        if isinstance(value, bool):
            return "true" if value else "false"
        return str(value)

    def element_by_id(self, raw_id: int) -> AccessibilityElement:
        """
        Find element by raw_id and return its properties for element finding.

        Args:
            raw_id: The raw_id to search for

        Returns:
            AccessibilityElement with backend_node_id set
        """
        # Get raw XML with raw_id attributes
        raw_xml = self.to_str()
        root = fromstring(f"<root>{raw_xml}</root>")

        # Find element with matching raw_id
        def find_element(elem: Element, target_id: str) -> Element | None:
            if elem.get("raw_id") == target_id:
                return elem
            for child in elem:
                result = find_element(child, target_id)
                if result is not None:
                    return result
            return None

        element = find_element(root, str(raw_id))
        if element is None:
            raise KeyError(f"No element with raw_id={raw_id} found")

        # Extract backendDOMNodeId for Chromium CDP nodes
        backend_node_id_str = element.get("backendDOMNodeId")
        if backend_node_id_str is None:
            raise ValueError(f"Element with raw_id={raw_id} has no backendDOMNodeId attribute")

        return AccessibilityElement(
            type=element.tag,
            backend_node_id=int(backend_node_id_str),
            frame=self._frame_map.get(raw_id),
            frame_chain=self._frame_chain_map.get(raw_id),
        )

    def scope_to_area(self, raw_id: int) -> "ChromiumAccessibilityTree":
        """Scope the tree to a smaller subtree identified by raw_id."""
        raw_xml = self.to_str()

        # Parse the XML (wrap in root if there are multiple top-level elements)
        try:
            root = fromstring(raw_xml)
            wrapped = False
        except Exception:
            root = fromstring(f"<root>{raw_xml}</root>")
            wrapped = True

        # Find the element with the matching raw_id
        def find_element(elem: Element, target_id: str) -> Element | None:
            if elem.get("raw_id") == target_id:
                return elem
            for child in elem:
                result = find_element(child, target_id)
                if result is not None:
                    return result
            return None

        search_root = root if not wrapped else root
        target_elem = find_element(search_root, str(raw_id))

        if target_elem is None:
            # If not found, return original tree
            return self

        # Convert the scoped element back to XML string
        indent(target_elem)
        scoped_xml = tostring(target_elem, encoding="unicode")

        return self._from_xml(scoped_xml, self._frame_map)
